SSA/Experimental/IntrinsicAsymptotics.lean

```diff
@@ -1,96 +1,72 @@
-  | beq : Op
-  | cst : ℕ → Op
-  deriving DecidableEq
+class OpSignature (Op : Type) (Ty : outParam (Type)) where
+  sig : Op → List Ty
+  outTy : Op → Ty
 
-def Op.sig : Op → List Ty
-  | .add => [.nat, .nat]
-  | .beq => [.nat, .nat]
-  | .cst _ => []
+class OpDenote (Op Ty : Type) [Goedel Ty] [OpSignature Op Ty] where
+  denote : (op : Op) → HVector toType (OpSignature.sig op) → (toType <| OpSignature.outTy op)
 
-def Op.outTy : Op → Ty
-  | .add => .nat
-  | .beq => .bool
-  | .cst _ => .nat
 
-inductive HVector (A : Ty → Type*) : List Ty → Type _
-  | nil : HVector A []
-  | cons {t : Ty} {l : List Ty} : A t → HVector A l → HVector A (t::l)
 
-/-- A very simple intrinsically typed expression. -/
-structure IExpr (Γ : Ctxt) (ty : Ty) : Type :=
-  (op : Op)
-  (ty_eq : ty = op.outTy)
-  (args : HVector (Ctxt.Var Γ) op.sig)
+/-
+  # Datastructures
+-/
 
-def IExpr.cst {Γ : Ctxt} (n : ℕ) : IExpr Γ .nat  :=
-  { op := Op.cst n
-    ty_eq := rfl
-    args := .nil }
+variable (Op : Type) {Ty : Type} [OpSignature Op Ty]
 
-def IExpr.add {Γ : Ctxt} (e₁ e₂ : Var Γ .nat) : IExpr Γ .nat :=
-  { op := Op.add
-    ty_eq := rfl
-    args := .cons e₁ <| .cons e₂ .nil }
+/-- A very simple intrinsically typed expression. -/
+structure IExpr (Γ : Ctxt Ty) (ty : Ty) : Type :=
+  (op : Op)
+  (ty_eq : ty = OpSignature.outTy op)
+  (args : HVector (Ctxt.Var Γ) <| OpSignature.sig op)
 
 /-- A very simple intrinsically typed program: a sequence of let bindings. -/
-inductive ICom : Ctxt → Ty → Type where
+inductive ICom : Ctxt Ty → Ty → Type where
   | ret (v : Γ.Var t) : ICom Γ t
-  | lete (e : IExpr Γ α) (body : ICom (Γ.snoc α) β) : ICom Γ β
+  | lete (e : IExpr Op Γ α) (body : ICom (Γ.snoc α) β) : ICom Γ β
 
-/-- `Lets Γ₁ Γ₂` is a sequence of lets which are well-formed under context `Γ₂` and result in
+/-- `Lets Op Γ₁ Γ₂` is a sequence of lets which are well-formed under context `Γ₂` and result in
     context `Γ₁`-/
-inductive Lets : Ctxt → Ctxt → Type where
-  | nil {Γ : Ctxt} : Lets Γ Γ
-  | lete (body : Lets Γ₁ Γ₂) (e : IExpr Γ₂ t) : Lets Γ₁ (Γ₂.snoc t)
+inductive Lets : Ctxt Ty → Ctxt Ty → Type where
+  | nil {Γ : Ctxt Ty} : Lets Γ Γ
+  | lete (body : Lets Γ₁ Γ₂) (e : IExpr Op Γ₂ t) : Lets Γ₁ (Γ₂.snoc t)
 
 
 
 /-
   # Definitions
 -/
 
-def HVector.map {A B : Ty → Type*} (f : ∀ (t : Ty), A t → B t) :
-    ∀ {l : List Ty}, HVector A l → HVector B l
-  | [], .nil => .nil
-  | t::_, .cons a as => .cons (f t a) (map f as)
-
-def HVector.foldl {A : Ty → Type*} {B : Type*} (f : ∀ (t : Ty), B → A t → B) :
-    ∀ {l : List Ty}, B → HVector A l → B
-  | [], b, .nil => b
-  | t::_, b, .cons a as => foldl f (f t b a) as
+variable {Op Ty : Type} [OpSignature Op Ty]
 
-@[reducible]
-def Op.denote : (op : Op) →
-    HVector Ty.toType op.sig → op.outTy.toType
-  | .cst n, _ => n
-  | .add, .cons a (.cons b .nil) => a + b
-  | .beq, .cons a (.cons b .nil) => a == b
+-- TODO: the following `variable` probably means we include these assumptions also in definitions
```

Comment:
no, I think variable is pretty smart about that. For example, if I
`#print Expr.denote` below here I get:
```
ef IExpr.denote : {Op Ty : Type} →
  [inst : OpSignature Op Ty] →
    [inst_1 : Goedel Ty] →
      [inst_2 : OpDenote Op Ty] → {Γ : Ctxt Ty} → {ty : Ty} → IExpr Op Γ ty → Ctxt.Valuation Γ → ⟦ty⟧ :=
fun {Op Ty} [OpSignature Op Ty] [Goedel Ty] [OpDenote Op Ty] {Γ} x x_1 x_2 =>
  match x, x_1, x_2 with
  | .(OpSignature.outTy op), { op := op, ty_eq := (_ : OpSignature.outTy op = OpSignature.outTy op), args := args },
    Γv => OpDenote.denote op (HVector.map (fun x v => Γv v) args)
    ```
    i.e. no `[DecidableEq Ty]` instance, even though it is using `Ty`.